### PR TITLE
HOSTEDCP-2166: Add ability to set Labels on HCP Pods

### PR DESCRIPTION
**Pull request by muraee**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Comment by muraee** [comment]
e2e will not pass until CNO/CSI changes are in place
CNO: https://github.com/openshift/cluster-network-operator/pull/2567

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 1]
```diff
@@ -610,6 +610,12 @@ type HostedClusterSpec struct {
 	//
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// labels when specified, define what custome labels are added to the hcp pods.
+	// Valid values are those in https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+	//
+	// +optional
```
Comment (enxebre) [inline api/hypershift/v1beta1/hostedcluster_types.go:617]: Please document what happens if you change this day 2.
Please include CEL validation for the allowed input here and include test cases for it within TestOnCreateAPIUX

File: test/e2e/create_cluster_test.go [hunk 2]
```diff
@@ -92,6 +92,61 @@ func TestOnCreateAPIUX(t *testing.T) {
...
+					expectedErrorSubstring string
+				}{
+					{
+						name: "when label key is empty it should fail",
+						mutateInput: func(hc *hyperv1.HostedCluster) {
+							hc.Spec.Labels = map[string]string{
+								"": "test-value",
+							}
+						},
+						expectedErrorSubstring: "label key must have two segments: an optional prefix and name, separated by a slash (/). The name segment is required and must be 63 characters or less, beginning and ending with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between. The prefix is optional. If specified, the prefix must be a DNS subdomain: a series of DNS labels separated by dots (.), not longer than 253 characters in total, followed by a slash (/)",
```
Comment (enxebre) [inline test/e2e/create_cluster_test.go:110]: nice, thanks!

File: api/hypershift/v1beta1/hosted_controlplane.go [hunk 3]
```diff
@@ -184,6 +184,15 @@ type HostedControlPlaneSpec struct {
...
+	// +kubebuilder:validation:XValidation:rule="self.all(key, size(self[key]) <= 63 && self[key].matches('^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$'))", message="label value must be 63 characters or less (can be empty), consist of alphanumeric characters, dashes (-), underscores (_) or dots (.), and begin and end with an alphanumeric character"
+	// +optional
```
Comment (enxebre) [inline api/hypershift/v1beta1/hosted_controlplane.go:194]: please include maxProperties

Comment (enxebre) [reply api/hypershift/v1beta1/hosted_controlplane.go:195]: e.g.
// +kubebuilder:validation:MaxProperties=10

Also what happens if I introduce the same label key more than once?
Can we please also check uniqueness?

Comment (muraee) [reply api/hypershift/v1beta1/hosted_controlplane.go:196]: > Also what happens if I introduce the same label key more than once?
> Can we please also check uniqueness?

this is a map, keys are implicitly validated to be unique, no?

Comment (muraee) [reply api/hypershift/v1beta1/hosted_controlplane.go:197]: updated the godoc

Comment (enxebre) [reply api/hypershift/v1beta1/hosted_controlplane.go:197]: So if duplicate entries are passed, latest one take prio? or something else?
Also what happens if the input passed is not valid? does this fail in a condition?

Comment (muraee) [reply api/hypershift/v1beta1/hosted_controlplane.go:197]: done

**Pull request merged**